js/data/bucket.js

```diff
@@ -0,0 +1,239 @@
+        if (secondElementBufferName && !buffers[secondElementBufferName]) {
+            buffers[secondElementBufferName] = createElementBuffer(shader.secondElementBufferComponents);
+        }
+    }
+
+    this.elementGroups = createElementGroups(this.shaders, this.buffers);
+};
+
+/**
+ * Get the name of the method used to add an item to a buffer.
+ * @param {string} shaderName The name of the shader that will use the buffer
+ * @param {string} type One of "vertex", "element", or "secondElement"
+ * @returns {string}
+ */
+Bucket.prototype.getAddMethodName = function(shaderName, type) {
+    return 'add' + capitalize(shaderName) + capitalize(type);
+};
+
+function createLayoutProperties(layer, zoom) {
+    var values = new StyleDeclarationSet('layout', layer.type, layer.layout, {}).values();
+    var fakeZoomHistory = { lastIntegerZoom: Infinity, lastIntegerZoomTime: 0, lastZoom: 0 };
+
+    var layout = {};
+    for (var k in values) {
+        layout[k] = values[k].calculate(zoom, fakeZoomHistory);
+    }
+
+    if (layer.type === 'symbol') {
+        // To reduce the number of labels that jump around when zooming we need
+        // to use a text-size value that is the same for all zoom levels.
+        // This calculates text-size at a high zoom level so that all tiles can
+        // use the same value when calculating anchor positions.
+        if (values['text-size']) {
+            layout['text-max-size'] = values['text-size'].calculate(18, fakeZoomHistory);
+            layout['text-size'] = values['text-size'].calculate(zoom + 1, fakeZoomHistory);
+        }
+        if (values['icon-size']) {
+            layout['icon-max-size'] = values['icon-size'].calculate(18, fakeZoomHistory);
+            layout['icon-size'] = values['icon-size'].calculate(zoom + 1, fakeZoomHistory);
+        }
+    }
+
+    return new LayoutProperties[layer.type](layout);
+}
+
+function createVertexAddMethod(shaderName, shader) {
+    if (!shader.vertexBuffer) return null;
+
+    // Find max arg length of all attribute value functions
+    var argCount = 0;
+    for (var i = 0; i < shader.attributes.length; i++) {
+        var attribute = shader.attributes[i];
+        argCount = Math.max(attribute.value.length, argCount);
+    }
+
+    var argIds = [];
+    for (var j = 0; j < argCount; j++) {
+        argIds.push('a' + j);
+    }
+
+    var body = '';
+    body += 'var attributes = this.shaders.' + shaderName + '.attributes;\n';
+    body += 'var elementGroups = this.elementGroups.' + shaderName + ';\n';
+    body += 'elementGroups.current.vertexLength++;\n';
+    body += 'return this.buffers.' + shader.vertexBuffer + '.push(\n';
+
+    for (var k = 0; k < shader.attributes.length; k++) {
+        body += '  attributes[' + k + '].value(' + argIds.join(', ') + ')';
+        body += (k !== shader.attributes.length - 1) ? ',\n' : '';
+    }
+    body += '\n) - elementGroups.current.vertexStartIndex;';
```

Comment:
We might have better luck with separate `push` calls for each separate value. There's was a severe performance regression on it in V8 that I reported https://code.google.com/p/v8/issues/detail?id=3883